Table without rows causes panic

I'm probably doing something silly, but `mdbook-pdf` and the regular `mdbook` html output both work.  Here is what I see:

```shell
    Replaced package `mdbook-pandoc v0.10.4` with `mdbook-pandoc v0.10.4` (executable `mdbook-pandoc`)
samitbasu@Mac code_standards % RUST_BACKTRACE=1 mdbook build               
2025-06-04 16:25:27 [INFO] (mdbook::book): Book building has started
2025-06-04 16:25:27 [INFO] (mdbook::book): Running the html backend
2025-06-04 16:25:28 [INFO] (mdbook::book): Running the pandoc backend
2025-06-04 16:25:28 [INFO] (mdbook::renderer): Invoking the "pandoc" renderer

thread 'main' panicked at /Users/samitbasu/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/mdbook-pandoc-0.10.4/src/preprocess/tree.rs:236:83:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0: _rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: core::option::unwrap_failed
   4: mdbook_pandoc::preprocess::tree::Emitter::serialize_node
   5: mdbook_pandoc::pandoc::native::SerializeList<W,mdbook_pandoc::pandoc::native::Block>::serialize_nested
   6: mdbook_pandoc::preprocess::tree::Emitter::emit
   7: mdbook_pandoc::pandoc::native::Serializer<W>::serialize
   8: mdbook_pandoc::preprocess::PreprocessChapter::preprocess
   9: <mdbook_pandoc::preprocess::Preprocess as core::iter::traits::iterator::Iterator>::next
  10: <mdbook_pandoc::Renderer as mdbook::renderer::Renderer>::render
  11: mdbook_pandoc::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
2025-06-04 16:25:28 [ERROR] (mdbook::renderer): Renderer exited with non-zero return code.
2025-06-04 16:25:28 [ERROR] (mdbook::utils): Error: Rendering failed
2025-06-04 16:25:28 [ERROR] (mdbook::utils): 	Caused By: The "pandoc" renderer failed
samitbasu@Mac code_standards % pandoc --version
pandoc 3.7.0.2
Features: +server +lua
Scripting engine: Lua 5.4
User data directory: /Users/samitbasu/.local/share/pandoc
Copyright (C) 2006-2024 John MacFarlane. Web: https://pandoc.org
This is free software; see the source for copying conditions. There is no
warranty, not even for merchantability or fitness for a particular purpose.
samitbasu@Mac code_standards % 
```

I installed `mdbook-pandoc` via `cargo install mdbook-pandoc --locked`.  I will try and dig around to see if I can figure out what is going wrong.

--- a/src/pandoc/native.rs
+++ b/src/pandoc/native.rs
@@ -390,6 +390,17 @@ impl SerializeElement for Text {
     }
 }
 
+impl SerializeElement for TableBody {
+    type Serializer<'a, 'book: 'a, 'p: 'a + 'book, W: io::Write + 'a> =
+        SerializeTableBody<'a, 'book, 'p, W>;
+
+    fn serializer<'a, 'book, 'p, W: io::Write>(
+        &mut self,
+        serializer: &'a mut Serializer<'p, 'book, W>,
+    ) -> Self::Serializer<'a, 'book, 'p, W> {
+        SerializeTableBody { serializer }
+    }
+}
 impl SerializeElement for Row {
     type Serializer<'a, 'book: 'a, 'p: 'a + 'book, W: io::Write + 'a> =
         SerializeRow<'a, 'book, 'p, W>;
@@ -483,6 +494,8 @@ pub struct Text;
 #[derive(Copy, Clone)]
 pub struct List<T>(T);
 #[derive(Copy, Clone)]
+pub struct TableBody;
+#[derive(Copy, Clone)]
 pub struct Row;
 #[derive(Copy, Clone)]
 pub struct Cell;
@@ -491,6 +504,7 @@ pub struct DefinitionListItem;
 
 pub type SerializeBlocks<'a, 'book, 'p, W> = SerializeList<'a, 'book, 'p, W, Block>;
 pub type SerializeInlines<'a, 'book, 'p, W> = SerializeList<'a, 'book, 'p, W, Inline>;
+pub type SerializeTableBodies<'a, 'book, 'p, W> = SerializeList<'a, 'book, 'p, W, TableBody>;
 pub type SerializeRows<'a, 'book, 'p, W> = SerializeList<'a, 'book, 'p, W, Row>;
 pub type SerializeCells<'a, 'book, 'p, W> = SerializeList<'a, 'book, 'p, W, Cell>;
 
@@ -516,6 +530,11 @@ pub struct SerializeText<'a, 'book, 'p, W: io::Write> {
 
 #[must_use]
 pub struct SerializeCode<'a, 'book, 'p, W: io::Write> {
+    serializer: &'a mut Serializer<'p, 'book, W>,
+}
+
+#[must_use]
+pub struct SerializeTableBody<'a, 'book, 'p, W: io::Write> {
     serializer: &'a mut Serializer<'p, 'book, W>,
 }
 
@@ -892,10 +911,7 @@ impl<'a, 'book, 'p, W: io::Write> SerializeBlock<'a, 'book, 'p, W> {
             impl Attributes,
             impl FnOnce(&mut SerializeRows<'_, 'book, 'p, W>) -> anyhow::Result<()>,
         ),
-        body: (
-            impl Attributes,
-            impl FnOnce(&mut SerializeRows<'_, 'book, 'p, W>) -> anyhow::Result<()>,
-        ),
+        bodies: impl FnOnce(&mut SerializeTableBodies<'_, 'book, 'p, W>) -> anyhow::Result<()>,
     ) -> anyhow::Result<()> {
         write!(self.serializer.unescaped(), "Table ")?;
 
@@ -939,14 +955,10 @@ impl<'a, 'book, 'p, W: io::Write> SerializeBlock<'a, 'book, 'p, W> {
 
         // Bodies: [TableBody Attr RowHeadColumns [Row] [Row]]
         {
-            let (attrs, rows) = body;
-            write!(self.serializer.unescaped(), " [(TableBody ")?;
-            self.serializer.write_attributes(attrs)?;
-            write!(self.serializer.unescaped(), " (RowHeadColumns 0) [] ")?;
-            let mut serializer = SerializeList::new(self.serializer, Row)?;
-            rows(&mut serializer)?;
+            write!(self.serializer.unescaped(), " ")?;
+            let mut serializer = SerializeList::new(self.serializer, TableBody)?;
+            bodies(&mut serializer)?;
             serializer.finish()?;
-            write!(self.serializer.unescaped(), ")]")?;
         }
 
         // Foot
@@ -984,6 +996,24 @@ impl<W: io::Write> SerializeAttribute<'_, '_, '_, W> {
         writer.write_all(val.as_bytes())?;
         writer.end_text()?;
         self.serializer.unescaped().write_all(b")")?;
+        Ok(())
+    }
+}
+
+impl<'book, 'p, W: io::Write> SerializeTableBody<'_, 'book, 'p, W> {
+    /// TableBody Attr RowHeadColumns [Row] [Row]
+    pub fn serialize_body(
+        &mut self,
+        attrs: impl Attributes,
+        rows: impl FnOnce(&mut SerializeRows<'_, 'book, 'p, W>) -> anyhow::Result<()>,
+    ) -> anyhow::Result<()> {
+        write!(self.serializer.unescaped(), "(TableBody ")?;
+        self.serializer.write_attributes(attrs)?;
+        write!(self.serializer.unescaped(), " (RowHeadColumns 0) [] ")?;
+        let mut serializer = SerializeList::new(self.serializer, Row)?;
+        rows(&mut serializer)?;
+        serializer.finish()?;
+        write!(self.serializer.unescaped(), ")")?;
         Ok(())
     }
 }

--- a/src/preprocess/tree.rs
+++ b/src/preprocess/tree.rs
@@ -233,7 +233,8 @@ impl<'book> Emitter<'book> {
                     let preprocessor = serializer.preprocessor();
                     let column_widths = preprocessor.column_widths(source);
                     let mut children = node.children();
-                    let (head, body) = (children.next().unwrap(), children.next().unwrap());
+                    let head = children.next().unwrap();
+                    let body = children.next();
                     debug_assert!(children.next().is_none());
 
                     let thead = match head.value() {
@@ -244,100 +245,88 @@ impl<'book> Emitter<'book> {
                         }
                         event => anyhow::bail!("expected table head, got {event:?}"),
                     };
-                    let tbody = match body.value() {
-                        Node::Element(Element::Html(element))
-                            if element.name.expanded() == expanded_name!(html "tbody") =>
-                        {
-                            element
-                        }
-                        event => anyhow::bail!("expected table body, got {event:?}"),
-                    };
+                    let body = body
+                        .map(|node| match node.value() {
+                            Node::Element(Element::Html(element))
+                                if element.name.expanded() == expanded_name!(html "tbody") =>
+                            {
+                                Ok((node, element))
+                            }
+                            event => anyhow::bail!("expected table body, got {event:?}"),
+                        })
+                        .transpose()?;
 
                     serializer.blocks()?.serialize_element()?.serialize_table(
                         (),
                         (alignment.iter().copied().map(Into::into)).zip(column_widths),
                         (&thead.attrs, |serializer| {
                             for row in head.children() {
-                                match row.value() {
-                                    Node::Element(Element::Html(element)) if element.name.expanded() == expanded_name!(html "tr") => {
-                                        serializer.serialize_element()?.serialize_row(&element.attrs, |cells| {
+                                let tr = match row.value() {
+                                    Node::Element(Element::Html(e)) if e.name.expanded() == expanded_name!(html "tr") => e,
+                                    event => anyhow::bail!("expected table row, got {event:?}"),
+                                };
+                                serializer.serialize_element()?.serialize_row(&tr.attrs, |cells| {
+                                    for cell in row.children() {
+                                        let th = match cell.value() {
+                                            Node::Element(Element::Html(e)) if e.name.expanded() == expanded_name!(html "th") => e,
+                                            event => anyhow::bail!("expected table cell, got {event:?}"),
+                                        };
+                                        for node in cell.children() {
+                                            cells.serialize_element()?.serialize_cell(
+                                                &th.attrs,
+                                                |blocks| {
+                                                    blocks.serialize_nested(|serializer| {
+                                                        self.serialize_node(node, serializer)
+                                                    })
+                                                },
+                                            )?;
+                                        }
+                                    }
+                                    Ok(())
+                                })?
+                            }
+                            Ok(())
+                        }),
+                        |serializer| {
+                            let Some((body, tbody)) = body else {
+                                return Ok(())
+                            };
+                            serializer.serialize_element()?.serialize_body(&tbody.attrs, |serializer| {
+                                for row in body.children() {
+                                    let tr = match row.value() {
+                                        Node::Element(Element::Html(e))
+                                            if e.name.expanded() == expanded_name!(html "tr") => e,
+                                        event => anyhow::bail!("expected table row, got {event:?}"),
+                                    };
+                                    serializer.serialize_element()?.serialize_row(
+                                        &tr.attrs,
+                                        |cells| {
                                             for cell in row.children() {
-                                                match cell.value() {
-                                                    Node::Element(Element::Html(element)) if element.name.expanded() == expanded_name!(html "th") => {
-                                                        for node in cell.children() {
-                                                            cells.serialize_element()?.serialize_cell(
-                                                                &element.attrs,
-                                                                |blocks| {
-                                                                    blocks.serialize_nested(|serializer| {
-                                                                        self.serialize_node(
-                                                                            node, serializer,
-                                                                        )
-                                                                    })
-                                                                },
-                                                            )?;
-                                                        }
-                                                    }
-                                                    event => {
-                                                        anyhow::bail!("expected table cell, got {event:?}")
-                                                    }
-                                                }
+                                                let td = match cell.value() {
+                                                    Node::Element(Element::Html(e))
+                                                        if e.name.expanded() == expanded_name!(html "td") => e,
+                                                    event => anyhow::bail!("expected table data (<td>), got {event:?}"),
+                                                };
+                                                cells
+                                                    .serialize_element()?
+                                                    .serialize_cell(&td.attrs, |blocks| {
+                                                        blocks.serialize_nested(
+                                                            |serializer| {
+                                                                for node in cell.children() {
+                                                                    self.serialize_node(node, serializer)?;
+                                                                }
+                                                                Ok(())
+                                                            },
+                                                        )
+                                                    })?
                                             }
                                             Ok(())
-                                        })?
-                                    }
-                                    event => anyhow::bail!("expected table row, got {event:?}"),
+                                        },
+                                    )?
                                 }
-                            }
-                            Ok(())
-                        }),
-                        (&tbody.attrs, |serializer| {
-                            for row in body.children() {
-                                match row.value() {
-                                    Node::Element(Element::Html(element))
-                                        if element.name.expanded() == expanded_name!(html "tr") =>
-                                    {
-                                        serializer.serialize_element()?.serialize_row(
-                                            &element.attrs,
-                                            |cells| {
-                                                for cell in row.children() {
-                                                    match cell.value() {
-                                                        Node::Element(Element::Html(element))
-                                                            if element.name.expanded()
-                                                                == expanded_name!(html "td") =>
-                                                        {
-                                                            cells
-                                                                .serialize_element()?
-                                                                .serialize_cell(&element.attrs, |blocks| {
-                                                                    blocks.serialize_nested(
-                                                                        |serializer| {
-                                                                            for node in
-                                                                                cell.children()
-                                                                            {
-                                                                                self.serialize_node(
-                                                                                    node, serializer,
-                                                                                )?;
-                                                                            }
-                                                                            Ok(())
-                                                                        },
-                                                                    )
-                                                                })?
-                                                        }
-                                                        event => {
-                                                            anyhow::bail!(
-                                                                "expected table data (<td>), got {event:?}"
-                                                            )
-                                                        }
-                                                    }
-                                                }
-                                                Ok(())
-                                            },
-                                        )?
-                                    }
-                                    event => anyhow::bail!("expected table row, got {event:?}"),
-                                }
-                            }
-                            Ok(())
-                        }),
+                                Ok(())
+                            })
+                        }
                     )
                 }
                 MdElement::FootnoteDefinition => Ok(()),

--- a/src/tests/tables.rs
+++ b/src/tests/tables.rs
@@ -3,7 +3,39 @@ use indoc::indoc;
 use super::{Chapter, Config, MDBook};
 
 #[test]
-fn tables() {
+fn empty() {
+    let book = MDBook::init()
+        .chapter(Chapter::new(
+            "",
+            indoc! {"
+                | Header1 | Header2 |
+                |---------|---------|
+            "},
+            "chapter.md",
+        ))
+        .config(Config::latex())
+        .build();
+    insta::assert_snapshot!(book, @r#"
+    ├─ log output
+    │  INFO mdbook::book: Running the pandoc backend    
+    │  INFO mdbook_pandoc::pandoc::renderer: Running pandoc    
+    │  INFO mdbook_pandoc::pandoc::renderer: Wrote output to book/latex/output.tex    
+    ├─ latex/output.tex
+    │ \begin{longtable}[]{@{}ll@{}}
+    │ \toprule\noalign{}
+    │ Header1 & Header2 \\
+    │ \midrule\noalign{}
+    │ \endhead
+    │ \bottomrule\noalign{}
+    │ \endlastfoot
+    │ \end{longtable}
+    ├─ latex/src/chapter.md
+    │ [Table ("", [], []) (Caption Nothing []) [(AlignDefault, ColWidthDefault), (AlignDefault, ColWidthDefault)] (TableHead ("", [], []) [Row ("", [], []) [Cell ("", [], []) AlignDefault (RowSpan 0) (ColSpan 0) [Plain [Str "Header1"]], Cell ("", [], []) AlignDefault (RowSpan 0) (ColSpan 0) [Plain [Str "Header2"]]]]) [(TableBody ("", [], []) (RowHeadColumns 0) [] [])] (TableFoot ("", [], []) [])]
+    "#);
+}
+
+#[test]
+fn basic() {
     let book = MDBook::init()
         .chapter(Chapter::new(
             "",
@@ -37,7 +69,7 @@ fn tables() {
 }
 
 #[test]
-fn wide_table() {
+fn wide() {
     let book = MDBook::init()
         .chapter(Chapter::new(
             "",

--- a/src/tests/tables.rs
+++ b/src/tests/tables.rs
@@ -30,7 +30,7 @@ fn empty() {
     │ \endlastfoot
     │ \end{longtable}
     ├─ latex/src/chapter.md
-    │ [Table ("", [], []) (Caption Nothing []) [(AlignDefault, ColWidthDefault), (AlignDefault, ColWidthDefault)] (TableHead ("", [], []) [Row ("", [], []) [Cell ("", [], []) AlignDefault (RowSpan 0) (ColSpan 0) [Plain [Str "Header1"]], Cell ("", [], []) AlignDefault (RowSpan 0) (ColSpan 0) [Plain [Str "Header2"]]]]) [(TableBody ("", [], []) (RowHeadColumns 0) [] [])] (TableFoot ("", [], []) [])]
+    │ [Table ("", [], []) (Caption Nothing []) [(AlignDefault, ColWidthDefault), (AlignDefault, ColWidthDefault)] (TableHead ("", [], []) [Row ("", [], []) [Cell ("", [], []) AlignDefault (RowSpan 0) (ColSpan 0) [Plain [Str "Header1"]], Cell ("", [], []) AlignDefault (RowSpan 0) (ColSpan 0) [Plain [Str "Header2"]]]]) [] (TableFoot ("", [], []) [])]
     "#);
 }
 

